sys/linux/init_images.go

```
@@ -0,0 +1,243 @@
+	newArgs := []prog.Arg{}
+	newSegments := []*mountImageSegment{}
+	for i, segment := range m.segments {
+		if filter(i, segment) {
+			newSegments = append(newSegments, segment)
+			newArgs = append(newArgs, m.segmentsGroup.Inner[i])
+		}
+	}
+	m.segments = newSegments
+	m.segmentsGroup.Inner = newArgs
+	m.segmentsCount.Val = uint64(len(newArgs))
+}
+
+type mountImageSegment struct {
+	data       *prog.DataArg
+	offset     *prog.ConstArg
+	size       *prog.ConstArg
+	parseError error
+}
+
+func parseImageSegment(segmentArg prog.Arg) *mountImageSegment {
+	ret := &mountImageSegment{}
+	segmentFields, ok := segmentArg.(*prog.GroupArg)
+	if segmentFields == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("it is not a group")}
+	}
+	if len(segmentFields.Inner) != 3 {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid number of nested fields")}
+	}
+	dataPtr, ok := segmentFields.Inner[0].(*prog.PointerArg)
+	if dataPtr == nil || dataPtr.Res == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid data field ptr")}
+	}
+	ret.data, ok = dataPtr.Res.(*prog.DataArg)
+	if ret.data == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid data arg")}
+	}
+	ret.size, ok = segmentFields.Inner[1].(*prog.ConstArg)
+	if ret.size == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid size arg")}
+	}
+	ret.offset, ok = segmentFields.Inner[2].(*prog.ConstArg)
+	if ret.offset == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid offset arg")}
+	}
+	return ret
+}
+
+func deactivateSyzMountImage(c *prog.Call) {
+	if len(c.Args) < 5 {
+		return
+	}
+	groupArg := c.Args[4]
+	newArg := groupArg.Type().DefaultArg(groupArg.Dir())
+	prog.RemoveArg(groupArg)
+	c.Args[4] = newArg
+}
+
+func parseSyzMountImage(c *prog.Call) (*mountImageArgs, error) {
+	if len(c.Args) < 5 {
+		return nil, fmt.Errorf("invalid number of arguments")
```

These errors are ignored in the end.
If something here will break in prod tomorrow, we won't know.
I think we need to panic here + extensively test.
Another option would be to save the error in the final image file, then users will come to us and complain about our bugs.

---

Now the code is saving errors to the files.